--- a/net/sched/sch_hfsc.c
+++ b/net/sched/sch_hfsc.c
@@ -176,12 +176,6 @@ struct hfsc_sched {
 
 #define	HT_INFINITY	0xffffffffffffffffULL	/* infinite time value */
 
-static bool cl_in_el_or_vttree(struct hfsc_class *cl)
-{
-	return ((cl->cl_flags & HFSC_FSC) && cl->cl_nactive) ||
-		((cl->cl_flags & HFSC_RSC) && !RB_EMPTY_NODE(&cl->el_node));
-}
-
 /*
  * eligible tree holds backlogged classes being sorted by their eligible times.
  * there is one eligible tree per hfsc instance.
@@ -1046,8 +1040,6 @@ hfsc_change_class(struct Qdisc *sch, u32
 	if (cl == NULL)
 		return -ENOBUFS;
 
-	RB_CLEAR_NODE(&cl->el_node);
-
 	err = tcf_block_get(&cl->block, &cl->filter_list, sch, extack);
 	if (err) {
 		kfree(cl);
@@ -1586,7 +1578,7 @@ hfsc_enqueue(struct sk_buff *skb, struct
 	sch->qstats.backlog += len;
 	sch->q.qlen++;
 
-	if (first && !cl_in_el_or_vttree(cl)) {
+	if (first && !cl->cl_nactive) {
 		if (cl->cl_flags & HFSC_RSC)
 			init_ed(cl, len);
 		if (cl->cl_flags & HFSC_FSC)
