--- a/net/sched/sch_hfsc.c	2025-07-17 10:48:13.714233578 +0300
+++ b/net/sched/sch_hfsc.c	2025-07-17 10:48:43.540041790 +0300
@@ -176,12 +176,6 @@
 
 #define	HT_INFINITY	0xffffffffffffffffULL	/* infinite time value */
 
-static bool cl_in_el_or_vttree(struct hfsc_class *cl)
-{
-	return ((cl->cl_flags & HFSC_FSC) && cl->cl_nactive) ||
-		((cl->cl_flags & HFSC_RSC) && !RB_EMPTY_NODE(&cl->el_node));
-}
-
 /*
  * eligible tree holds backlogged classes being sorted by their eligible times.
  * there is one eligible tree per hfsc instance.
@@ -1043,8 +1037,6 @@
 	if (cl == NULL)
 		return -ENOBUFS;
 
-	RB_CLEAR_NODE(&cl->el_node);
-
 	err = tcf_block_get(&cl->block, &cl->filter_list, sch, extack);
 	if (err) {
 		kfree(cl);
@@ -1582,7 +1574,7 @@
 	sch->qstats.backlog += len;
 	sch->q.qlen++;
 
-	if (first && !cl_in_el_or_vttree(cl)) {
+	if (first && !cl->cl_nactive) {
 		if (cl->cl_flags & HFSC_RSC)
 			init_ed(cl, len);
 		if (cl->cl_flags & HFSC_FSC)
