Index: linux-5.15.197/net/sched/act_connmark.c
===================================================================
--- linux-5.15.197.orig/net/sched/act_connmark.c
+++ linux-5.15.197/net/sched/act_connmark.c
@@ -234,6 +234,23 @@ static void tcf_connmark_cleanup(struct
 		kfree_rcu(parms, rcu);
 }
 
+static int tcf_connmark_walker(struct net *net, struct sk_buff *skb,
+			       struct netlink_callback *cb, int type,
+			       const struct tc_action_ops *ops,
+			       struct netlink_ext_ack *extack)
+{
+	struct tc_action_net *tn = net_generic(net, act_connmark_ops.net_id);
+
+	return tcf_generic_walker(tn, skb, cb, type, ops, extack);
+}
+
+static int tcf_connmark_search(struct net *net, struct tc_action **a, u32 index)
+{
+	struct tc_action_net *tn = net_generic(net, act_connmark_ops.net_id);
+
+	return tcf_idr_search(tn, a, index);
+}
+
 static struct tc_action_ops act_connmark_ops = {
 	.kind		=	"connmark",
 	.id		=	TCA_ID_CONNMARK,
@@ -242,6 +259,8 @@ static struct tc_action_ops act_connmark
 	.dump		=	tcf_connmark_dump,
 	.init		=	tcf_connmark_init,
 	.cleanup	=	tcf_connmark_cleanup,
+	.walk		=	tcf_connmark_walker,
+	.lookup		=	tcf_connmark_search,
 	.size		=	sizeof(struct tcf_connmark_info),
 };
 
